src/Bicep.Core/Analyzers/Linter/Rules/MaxNumberVariablesRule.cs

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Semantics;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Bicep.Core.Analyzers.Linter.Rules
+{
+    public sealed class MaxNumberVariablesRule : LinterRuleBase
+    {
+        public new const string Code = "max-variables";
+        public MaxNumberVariablesRule() : base(
+            code: Code,
+            description: CoreResources.MaxNumberVariablesRuleDescription,
+            docUri: new Uri($"https://aka.ms/bicep/linter/{Code}"),
+            diagnosticLabel: Diagnostics.DiagnosticLabel.Unnecessary,
+            diagnosticLevel: DiagnosticLevel.Error)
+        { }
+
+        public override string FormatMessage(params object[] values)
+        {
+            return string.Format(CoreResources.MaxNumberVariablesRuleMessageFormat, values);
+        }
+
+        override public IEnumerable<IDiagnostic> AnalyzeInternal(SemanticModel model)
+        {
+            if (model.Root.VariableDeclarations.Count() > 256) {
+                return model.Root.VariableDeclarations.Select(param => CreateDiagnosticForSpan(param.NameSyntax.Span, param.Name));
```

Comment:
>param.NameSyntax.Span [](http://example.com/codeflow?start=95&length=21)

Right now, this puts the diagnostic on every variable declaration when you exceed the limit. Should we put it the diagnostic on the first one? Or only on the ones that cause us to go above the limit?

Reply:
I started from the fact that you can be typing that var anywhere, and that you want to get a squiggly line on that var you just added. But it gives a ton of output that might mask other rule results. 

Thinking about it again, it would make more sense to just have this on the first element. That would give us only a single message where we can still say "you have <number> vars. Maximum allowed vars is <max>".

What do you think?

Reply:
One more question on this, should we mark the full declaration or only the value of the name property? Currently I only mark the name property.

this:
![image](https://user-images.githubusercontent.com/1624852/151664054-38588c0f-36a9-426f-8127-38d5de8ffbac.png)
or
![image](https://user-images.githubusercontent.com/1624852/151664343-75e0d8eb-6ee6-4af7-9839-e89c8a01bb01.png)


or for variables, this:
![image](https://user-images.githubusercontent.com/1624852/151664165-8a50bcde-f04a-42ca-9441-b34fbf6dcf57.png)
or
![image](https://user-images.githubusercontent.com/1624852/151664182-5e4f371b-2e28-4b1b-8601-d376b7c93866.png)

Reply:
Marking the first var sounds good to me.

> One more question on this, should we mark the full declaration or only the value of the name property? Currently I only mark the name property.

Personally, I prefer showing squiggly line under the symbolic names, even for resources. Marking the name property makes me think that the name itself is invalid.

Reply:
Agree

Reply:
Done.
![image](https://user-images.githubusercontent.com/1624852/152029761-c948c7e8-dcf2-4d5f-a169-d8792a700cd4.png)